File: notebookspython/04-palindrome_permutation.py
class Solution:
    
    def canPermutePalindrome(self, s: str) -> bool:
        s = s.replace(" ", "").lower()
        
        odd_chars = set()
    
        # 遍历字符串，利用集合对消
        for char in s:
            if char in odd_chars:
                odd_chars.remove(char)
            else:
                odd_chars.add(char)
                
        # 直接返回判断结果（函数结束）
        return len(odd_chars) <= 1

    # 一个简单的密码强度初步检测示例
    def is_password_too_simple(password: str) -> bool:
        if len(password) < 8:
            return True
        
    # 字符太单一（利用集合去重）
        if len(set(password)) < 4:
            return True
        
        return False
    
if __name__ == "__main__":
    # 创建类的实例
    sol = Solution()
    
    # 测试用例 1
    test1 = "code"
    print(f"'{test1}' 是否可以排成回文？", sol.canPermutePalindrome(test1))  # 预期输出: False
    
    # 测试用例 2
    test2 = "tactcoa"
    print(f"'{test2}' 是否可以排成回文？", sol.canPermutePalindrome(test2))  # 预期输出: True

File: notebooksc#/04-palindrome-permutation.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

public class Solution04PalindromePermutation
{
    // Solution: Using a set to track odd character counts
    public bool CanPermutePalindrome(string s)
    {
        s = s.Replace(" ", "").ToLower();
        
        HashSet<char> oddChars = new HashSet<char>();

        // Traverse the string and use a set to track odd counts
        foreach (char c in s)
        {
            if (oddChars.Contains(c))
                oddChars.Remove(c);
            else
                oddChars.Add(c);
        }

        // A string can be rearranged into a palindrome if at most one character has an odd count
        return oddChars.Count <= 1;
    }

    static void Main()
    {
        var sol = new Solution04PalindromePermutation();

        // Test case 1
        string test1 = "code";
        bool result1 = sol.CanPermutePalindrome(test1);
        Console.WriteLine($"'{test1}' can be rearranged to a palindrome? {result1}"); // Expected: False

        // Test case 2
        string test2 = "tactcoa";
        bool result2 = sol.CanPermutePalindrome(test2);
        Console.WriteLine($"'{test2}' can be rearranged to a palindrome? {result2}"); // Expected: True

        // Additional test cases
        Debug.Assert(sol.CanPermutePalindrome("code") == false);
        Debug.Assert(sol.CanPermutePalindrome("tactcoa") == true);
        Debug.Assert(sol.CanPermutePalindrome("a") == true);
        Debug.Assert(sol.CanPermutePalindrome("ab") == false);
        Debug.Assert(sol.CanPermutePalindrome("aab") == true);

        Console.WriteLine("All test cases passed ✅");
    }
}
